memory_w25q1128jv_spi/memory_w25q1128jv_spi.cc

```diff
@@ -0,0 +1,220 @@
+    HAL_GPIO_WritePin(_csPort, _csPin, GPIO_PIN_SET);
+
+    // input instruction hex + 3 dummy bytes
+    uint8_t DEVICE_ID[4] = {0x90, 0x00, 0x00, 0x00};
+    // output array of bytes
+    uint8_t deviceValue[2];
+
+    // SPI protocol
+    HAL_GPIO_WritePin(_csPort, _csPin, GPIO_PIN_RESET);
+    if (HAL_SPI_Transmit(_hspi, DEVICE_ID, 4, SPI_Timeout) != HAL_OK) return -1;
+    if (HAL_SPI_Receive(_hspi, deviceValue, 2, SPI_Timeout) != HAL_OK) return -1;
+    HAL_GPIO_WritePin(_csPort, _csPin, GPIO_PIN_SET);
+
+    // returns the 2 bytes as a uint16_t
+    return (int)(deviceValue[0] << 8 | deviceValue[1]);
+}
+
+int MemoryW25q1128jvSpi::ReadStatusReg1() {
+    // Control pins his for no false memory actions
+    HAL_GPIO_WritePin(_wpPort, _wpPin, GPIO_PIN_SET);
+    HAL_GPIO_WritePin(_holdPort, _holdPin, GPIO_PIN_SET);
+    HAL_GPIO_WritePin(_csPort, _csPin, GPIO_PIN_SET);
+
+    // input operation code for reading status register 1
+    uint8_t READ_STATUS_REG_1[1] = {0x05};
+    // output -- 8 bit value of the status register
+    uint8_t regValue[1];
+
+    HAL_GPIO_WritePin(_csPort, _csPin, GPIO_PIN_RESET);
+    if (HAL_SPI_Transmit(_hspi, READ_STATUS_REG_1, 1, SPI_Timeout) != HAL_OK) return -1;
+    if (HAL_SPI_Receive(_hspi, regValue, 1, SPI_Timeout) != HAL_OK) return -1;
+    HAL_GPIO_WritePin(_csPort, _csPin, GPIO_PIN_SET);
+
+    return (int)regValue[0];
+}
+
+int MemoryW25q1128jvSpi::ChipErase() {
+    // Control pins his for no false memory actions
+    HAL_GPIO_WritePin(_wpPort, _wpPin, GPIO_PIN_SET);
+    HAL_GPIO_WritePin(_holdPort, _holdPin, GPIO_PIN_SET);
+    HAL_GPIO_WritePin(_csPort, _csPin, GPIO_PIN_SET);
+
+    /* Set Write Enable to 1 */
+    uint8_t CHIP_ENABLE[1] = {0x06};
+
+    HAL_GPIO_WritePin(_csPort, _csPin, GPIO_PIN_RESET);
+    if (HAL_SPI_Transmit(_hspi, CHIP_ENABLE, 1, SPI_Timeout) != HAL_OK) return -1;
+    HAL_GPIO_WritePin(_csPort, _csPin, GPIO_PIN_SET);
+    ;
+    while ((MemoryW25q1128jvSpi::ReadStatusReg1() & 0x02) != 2) {
+    }
+
+    /* Initiate Chip Erase */
+    uint8_t CHIP_ERASE[1] = {0xC7};
+
+    HAL_GPIO_WritePin(_csPort, _csPin, GPIO_PIN_RESET);
+    if (HAL_SPI_Transmit(_hspi, CHIP_ERASE, 1, SPI_Timeout) != HAL_OK) return -1;
+    HAL_GPIO_WritePin(_csPort, _csPin, GPIO_PIN_SET);
+
+    // Waits until busy bit is = 0
+    // write busy is 1st bit in status reg
+    while ((MemoryW25q1128jvSpi::ReadStatusReg1() & 0x01) != 0) {
+    }
+
+    return 0;
+}
+
+MemoryW25q1128jvSpi::State MemoryW25q1128jvSpi::ChipWrite(uint8_t (&data)[64]) {
+    // Control pins his for no false memory actions
+    HAL_GPIO_WritePin(_wpPort, _wpPin, GPIO_PIN_SET);
+    HAL_GPIO_WritePin(_holdPort, _holdPin, GPIO_PIN_SET);
+    HAL_GPIO_WritePin(_csPort, _csPin, GPIO_PIN_SET);
+
+    // if address variable goes over the amount of pages memory has
+    // no more collected data will be written into the memory
+    if (address > 0xFFFFFF) _state = State::ERROR;
```

Comment:
i think just `_state = ERROR;` is fine, but we can discuss this further

Reply:
yea, I remember looking at your altimeter drivers and seeing you doing the `_state = ERROR;`. I honestly don't know why the compiler isn't telling you that the syntax is an error because it was for me. I asked Eric about it, can't remember what he said but he's also confused why yours works